From b6715720f22b0306a9e2975b6b294a240bd1933f Mon Sep 17 00:00:00 2001
From: "penikelapati.kumar" <penikelapati.kumar@lge.com>
Date: Wed, 17 Feb 2021 14:12:35 +0530
Subject: [PATCH] Fix crash when wifi tethering set false

:Release Notes:
Fix connman crash when calling wifi tethering

:Detailed Notes:
Fix crash when wifi set tethering false.
FIx crash when p2p/addservice and p2p/settethering.

:Testing Performed:
Build successfully.
Ethrenet/wifi connection successfully.

:QA Notes:
N/A

:Issues Addressed:
[PLAT-137131] [webOS OSE]: Failed to enable tethering mode
---
 plugins/wifi.c  | 2 +-
 src/group.c     | 5 ++---
 src/p2pgo.c     | 2 ++
 src/tethering.c | 3 ++-
 4 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/plugins/wifi.c b/plugins/wifi.c
index 0945a4a..e1093ed 100644
--- a/plugins/wifi.c
+++ b/plugins/wifi.c
@@ -1504,7 +1504,7 @@ static void register_peer_service_cb(int result,
 
 	DBG("");
 
-	if (reg_data->callback)
+	if (reg_data && reg_data->callback)
 		reg_data->callback(result, reg_data->user_data);
 
 	g_free(reg_data);
diff --git a/src/group.c b/src/group.c
index 83661cf..e55358b 100644
--- a/src/group.c
+++ b/src/group.c
@@ -85,11 +85,9 @@ static int set_tethering(struct connman_group *group,
 			&val);
 
 	if (enabled == TRUE) {
-		__connman_p2p_go_set_enabled();
 		__connman_p2p_go_tethering_set_enabled();
 	} else {
 		__connman_p2p_go_tethering_set_disabled();
-		__connman_p2p_go_set_disabled();
 	}
 
 	return 0;
@@ -554,7 +552,8 @@ static int group_register(struct connman_group *group)
 								group_methods, group_signals,
 								NULL, group, NULL);
 
-	__connman_p2p_set_dhcp_pool(NULL);
+	if (!group->autonomous)
+		__connman_p2p_set_dhcp_pool(NULL);
 	group_added_signal(group);
 
 	return 0;
diff --git a/src/p2pgo.c b/src/p2pgo.c
index e9d5f2d..5c0b7f7 100644
--- a/src/p2pgo.c
+++ b/src/p2pgo.c
@@ -237,6 +237,8 @@ void __connman_p2p_go_tethering_set_enabled(void)
 	const char *subnet_mask;
 	const char *start_ip;
 
+	if (!dhcp_ippool)
+		return;
 	subnet_mask = __connman_ippool_get_subnet_mask(dhcp_ippool);
 	start_ip = __connman_ippool_get_start_ip(dhcp_ippool);
 
diff --git a/src/tethering.c b/src/tethering.c
index 5f371e7..392ec18 100644
--- a/src/tethering.c
+++ b/src/tethering.c
@@ -855,8 +855,9 @@ void __connman_tethering_client_register(const char *addr)
 
 void __connman_tethering_client_unregister(const char *addr)
 {
+	char *remove_addr = g_strdup(addr);
 	g_hash_table_remove(clients_table, addr);
-	client_removed(addr);
+	client_removed(remove_addr);
 }
 
 int __connman_tethering_init(void)
-- 
2.26.0

